hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/file/OMRecoverLeaseRequest.java

```diff
@@ -0,0 +1,278 @@
+          volumeName, bucketName);
+
+      validateBucketAndVolume(omMetadataManager, volumeName, bucketName);
+
+      String openKeyEntryName = doWork(ozoneManager, recoverLeaseRequest,
+          transactionLogIndex);
+
+      // Prepare response
+      boolean responseCode = true;
+      omResponse.setRecoverLeaseResponse(RecoverLeaseResponse.newBuilder()
+              .setResponse(responseCode).build())
+          .setCmdType(RecoverLease);
+      omClientResponse =
+          new OMRecoverLeaseResponse(omResponse.build(), getBucketLayout(),
+              openKeyEntryName);
+      omMetrics.incNumRecoverLease();
+      LOG.debug("Key recovered. Volume:{}, Bucket:{}, Key:{}", volumeName,
+          bucketName, keyName);
+    } catch (IOException ex) {
+      LOG.error("Fail for recovering lease. Volume:{}, Bucket:{}, Key:{}",
+          volumeName, bucketName, keyName, ex);
+      exception = ex;
+      omMetrics.incNumRecoverLeaseFails();
+      omResponse.setCmdType(RecoverLease);
+      omClientResponse = new OMRecoverLeaseResponse(
+          createErrorOMResponse(omResponse, ex), getBucketLayout());
+    } finally {
+      if (omClientResponse != null) {
+        omClientResponse.setFlushFuture(
+            ozoneManagerDoubleBufferHelper.add(omClientResponse,
+                transactionLogIndex));
+      }
+      if (acquiredLock) {
+        omMetadataManager.getLock().releaseWriteLock(BUCKET_LOCK, volumeName,
+            bucketName);
+      }
+    }
+
+    // Audit Log outside the lock
+    auditLog(ozoneManager.getAuditLogger(), buildAuditMessage(
+        OMAction.RECOVER_LEASE, auditMap, exception,
+        getOmRequest().getUserInfo()));
+
+    return omClientResponse;
+  }
+
+  private String doWork(OzoneManager ozoneManager,
+      RecoverLeaseRequest recoverLeaseRequest, long transactionLogIndex)
+      throws IOException {
+
+    final long volumeId = omMetadataManager.getVolumeId(volumeName);
+    final long bucketId = omMetadataManager.getBucketId(
+        volumeName, bucketName);
+    Iterator<Path> pathComponents = Paths.get(keyName).iterator();
+    long parentID = OMFileRequest.getParentID(volumeId, bucketId,
+        pathComponents, keyName, omMetadataManager,
+        "Cannot recover file : " + keyName
+            + " as parent directory doesn't exist");
+    String fileName = OzoneFSUtils.getFileName(keyName);
+    String dbFileKey = omMetadataManager.getOzonePathKey(volumeId, bucketId,
+        parentID, fileName);
+
+    OmKeyInfo keyInfo = getKey(dbFileKey);
+    if (keyInfo == null) {
+      throw new OMException("Key:" + keyName + " not found", KEY_NOT_FOUND);
+    }
+    String openKeyEntryName = getOpenFileEntryName(volumeId, bucketId,
+        parentID, fileName);
+    if (openKeyEntryName != null) {
+      checkFileState(keyInfo, openKeyEntryName);
+      commitKey(dbFileKey, keyInfo, ozoneManager, transactionLogIndex);
+      removeOpenKey(openKeyEntryName, transactionLogIndex);
+    }
```

Comment:
Get `clientId` from metedata, i.e.
```java
    OmKeyInfo keyInfo = omMetadataManager.getKeyTable(getBucketLayout()).get(dbFileKey);
    if (keyInfo == null) {
      throw new OMException("Key:" + keyName + " not found", KEY_NOT_FOUND);
    }
    final String clientId = keyInfo.getMetadata().get(
        OzoneConsts.HSYNC_CLIENT_ID);
    if (clientId == null) {
      throw new OMException("Key:" + keyName + " is closed", KEY_ALREADY_EXISTS);
    }
    final String openKeyEntryName = omMetadataManager.getOpenFileName(
        volumeId, bucketId, parentID, fileName, Long.parseLong(clientId));

    checkFileState(keyInfo, openKeyEntryName);
    commitKey(dbFileKey, keyInfo, ozoneManager, transactionLogIndex);
    removeOpenKey(openKeyEntryName, transactionLogIndex, omMetadataManager);

    return openKeyEntryName;
```

Reply:
Actually now that I run this through TestOMRecoverLeaseRequest, I realized that this is not the same as HDFS. In HDFS, closing a closed file is allowed, essentially a no-op. HBase WAL writer actually calls recover lease before it starts, even if the WAL file is closed. So adding this check would break HBase.

What if we just log a warning instead?

Reply:
Sure.